### ISSUE TITLE
Useability and performance improvement

### ISSUE THREAD

**Issue by hans-schmidt** [RavenProject/Ravencoin#309]
**I would like to propose the addition of something like the following:**

A) RPC call burnsubasset  "subasset_name" "address" qty
* works only if you have the key for "address" and the subassetcontrol is also present at "address"
* implemented by creating a "R_subasset_burnxxxx" address which the client understands as a burn address

B) RPC call burnsubassetcontrol "subasset_name" "address"
* works only if you have the key for "address"
* implemented by creating a "R_subasset_control_burnxxxx" address which the client understands as a burn address

C) RPC call burnasset  "asset_name" "address" qty
* works only if you have the key for "address"
* implemented by creating a "R_asset_burnxxxx" address which the client understands as a burn address

D) RPC call burnassetcontrol "asset_name" "address"
* works only if you have the key for "address" and there are no assets or subassets in existance at any address except for the assetcontrol which is at "address"
* implemented by creating a "R_asset_control_burnxxxx" address which the client understands as a burn address

E) RPC call replacesubassetcontrol "old_subasset_name" "new_subasset_name" "address" ( units ) ( reissuable ) ( has_ipfs ) "( ipfs_hash )"
* works only if you have the key for "address" and there are tokens of old_subasset_name in existence at any address except for old_subasset_name's assetcontrol which is at "address"
* implemented by doing a "burnsubassetcontrol old_subasset_name" followed by a "issue new_subasset_name" but with the 100 RVN burn waived

F) RPC call replaceassetcontrol "old_asset_name" "new_asset_name" "address" ( units ) ( reissuable ) ( has_ipfs ) "( ipfs_hash )"
* works only if you have the key for "address" and there are no token assets or subassets of old_asset_name in existance at any address except for old_asset_name's assetcontrol which is at "address"
* implemented by doing a "burnassetcontrol old_subasset_name" followed by a "issue new_asset_name" but with the 500 RVN burn waived

**RPC calls A,B,C,D & E can be used to improve the blockchain performance, reduce user liability, and reduce unwanted display clutter**

Users who have assets they don't want can burn them. Because the node client understands the burn addresses, these asset UTXOs no longer need to remain in the asset LevelDB
* can be used to "clean up" the user display as well as the asset LevelDB if the user created them but no longer wants them
* can be used to "clean up" the user display as well as the asset LevelDB if the user received them as spam and they are objectionable (ie ipfs links to illegal material)

**RPC calls A,B,C & D together can also be used in preparation for using RPC call E: "replaceassetcontrol"**

**RPC call E would vastly improve the suitability of Ravencoin for a variety of applications, could dramatically improve blockchain performance, and would increase user's willingness to spend large amounts of money to purchase assets by insuring that they would never become worthless.**

The following application scenarios provide further insight:

>   1) If I run a concert ticketing service, I might buy the subasset called "PINK_FLOYD_2018_BOSTON" whose tokens I sell and can be traded. After the concert is over, I would like to re-use that subasset with the name of a new concert. But I can't. I have to buy new subassets. The old subassets are useless, and for the rest of eternity every node on the Ravencoin network has to keep track in its assets LevelDB of the existence of those and the growing list of all other useless subassets. It would be better if I could recollect all the subasset tokens into one address, burn them, and then do a "replacesubassetcontrol" with a new name.

>   2) Similarly, on Oct 31st, I might spend $10 to buy the "BEANIE_BABY" asset name. On Dec 1st, with RVN at $10, that asset name is theoretically worth $5000. But unfortunately, nobody wants Beanie Babies (duh!) and nobody wants the name. So that asset name is actually worth nothing and again all nodes on the Ravencoin network have to keep track of that forever. Meanwhile, my friend thinks that there is a market for something called "CRYPTO_KITTIES" but he is unsure whether he wants to risk $5000 on the name. The market would be better served if I could sell my asset name to my friend for $1000. I could recollect all my asset and subasset tokens into one address, burn them all, sell the assetcontrol to my friend, and he could "replaceassetcontrol" with a new name.

There will be cases where you fail to be able to do these operation because you can't get all the tokens back. But depending on how you used and distributed the tokens, there will be many cases where you CAN get them all back and then these operations will be helpful. Also, by using multisig or other fancy scripts there may be good bitcoin-compliant ways to help applications get the tokens back.

**Comment by TronBlack** [comment]
I agree with the burn.  It is easy, if we standardize on a burn address, to optimize the UTXO by not storing the UTXO set for burned assets.

I like the conditions you've added.   It is currently possible to burn any asset, sub-asset, or control asset by sending to the burn address.   The RPC calls would be just convenience functions for the users, but may be beneficial. The UTXO optimization can be added at any time.

I don't think renaming assets is a good idea for a number of reasons, including that there should be some risk for reserving asset names on-chain.

Philosophically, the user is in control of the asset, so there will not be a way to "recall" or "get the tokens back".  The user can always burn their own tokens.

**Comment by hans-schmidt** [comment]
@TronBlack Thanks for the feedback. I share your concerns about renaming/replacing. I included it because being able to reuse assets may also help mitigate some of the concerns in the burn-discussion by not wasting assets. Also, I am concerned that the asset UTXO will grow very quickly, and asset reuse is the primary motivator which users would have for burning assets with names which are no longer useful. Without incentive, users may not bother to burn.

Unfortunately, anything which is free of cost gets abused. Would you feel differently about renaming/replacing if it cost a fraction of the "issue asset" cost instead of zero cost?

Human-readable asset names are one of RVN's nice features, But at an application level, an application can always "rename" a token before showing it to the user.
So if you feel strongly about not wanting renaming/replacing, I won't puruse it any further.

**Comment by cfox** [comment]
Tron:

> I agree with the burn. It is easy
